File: src/data_wrangling/feature_engineering.R
# ---- Risk scores ----
#' Compute HATCH score
#'
#' Computes the HATCH score for each patient based on their age and
#' comorbidities.
#'
#' @param df A data frame containing patient data with columns for age,
#' hypertension, OSA, COPD, stroke, and heart failure.
#'
#' @return A data frame with an additional column for the HATCH score.
compute_hatch_score <- function(df) {
    df_result <- df |>
        dplyr::mutate(
            hatch_score =
                dplyr::if_else(age > 75, 1, 0) +
                    dplyr::if_else(hypertension == "yes", 1, 0) +
                    dplyr::if_else(COPD == "yes", 1, 0) +
                    dplyr::if_else(stroke == "yes", 2, 0) +
                    dplyr::if_else(heart_failure == "yes", 2, 0)
        )

    return(df_result)
}

#' Compute CHADS2 score
#'
#' Computes the CHADS2 score for each patient based on their age and
#' comorbidities.
#'
#' @param df A data frame containing patient data with columns for age,
#' hypertension, diabetes, stroke, stroke, and heart failure.
#'
#' @return A data frame with an additional column for the CHADS2 score.
compute_chads2_score <- function(df) {
    df_result <- df |>
        dplyr::mutate(
            chads2_score =
                dplyr::if_else(age >= 75, 1, 0) +
                    dplyr::if_else(hypertension == "yes", 1, 0) +
                    dplyr::if_else(diabetes == "yes", 1, 0) +
                    dplyr::if_else(stroke == "yes", 2, 0) +
                    dplyr::if_else(heart_failure == "yes", 1, 0)
        )

    return(df_result)
}

#' Compute BASE-AF2 score
#'
#' Computes the BASE-AF2 score for each patient based on their comorbidities.
#'
#' @param df A data frame containing patient data with columns for BMI,
#' atrial enlargement, smoking, ERAF, AF duration and AF type
#'
#' @return A data frame with an additional column for the BASE-AF2 score.
compute_base_af2_score <- function(df) {
    df_result <- df |>
        dplyr::mutate(
            base_af2_score =
                dplyr::if_else(BMI >= 28, 1, 0) +
                    dplyr::if_else((sex == "male" & LA_enlargment %in% c("moderate", "severe")) | (sex == "female" & LA_enlargment %in% c("mild", "moderate", "severe")), 1, 0) +
                    dplyr::if_else(smoking_status == "yes", 1, 0) +
                    dplyr::if_else(ERAF == "yes", 1, 0) +
                    dplyr::if_else(AF_duration > 6, 1, 0) +
                    dplyr::if_else(AF_type == "persistent", 1, 0)
        )
    return(df_result)
}

#' Compute MB-LATER score
#'
#' Computes the MB-LATER score for each patient based on their comorbidities.
#'
#' @param df A data frame containing patient data with columns for sex,
#' bundle branch block, atrial enlargement, AF type and ERAF.
#'
#' @return A data frame with an additional column for the MB-LATER score.
compute_mb_later_score <- function(df) {
    df_result <- df |>
        dplyr::mutate(
            mb_later_score = dplyr::if_else(sex == "male", 1, 0) +
                dplyr::if_else(bundle_branch_R == "yes" | bundle_branch_L == "yes", 1, 0) +
                dplyr::if_else(sex == "female" & LA_enlargment %in% c("moderate", "severe"), 1, 0) +
                dplyr::if_else(sex == "male" & LA_enlargment == "severe", 1, 0) +
                dplyr::if_else(AF_type == "persistent", 1, 0) +
                dplyr::if_else(ERAF == "yes", 1, 0)
        )
    return(df_result)
}

# ---- New features ----
#' Compute metabolic syndrome
#'
#' Computes the metabolic syndrome for each patient based on their
#' comorbidities.
#'
#' @param df A data frame containing patient data with columns for central
#' BMI, triglycerides, HDL, hypertension and glucose.
#'
#' @return A data frame with an additional column for the metabolic syndrome.
compute_metabolic_syndrome <- function(df) {
    df_result <- df |>
        dplyr::mutate(
            mets_obesity = dplyr::if_else(BMI >= 30, 1, 0, missing = 0),
            mets_triglycerides = dplyr::if_else(
                triglycerides >= 150 | hypolipidemic_meds == "yes",
                1, 0, missing = 0
            ),
            mets_hdl = dplyr::if_else(
                (sex == "male" & HDL < 40) |
                    (sex == "female" & HDL < 50),
                1, 0, missing = 0
            ),
            mets_hypertension = dplyr::if_else(
                hypertension == "yes", 1, 0, missing = 0
            ),
            mets_glucose = dplyr::if_else(
                glucose >= 100 | diabetes == "yes", 1, 0, missing = 0
            ),

            # Sum up comorbidities
            mets_sum = mets_obesity + mets_triglycerides + mets_hdl + mets_hypertension + mets_glucose,
            metabolic_syndrome = dplyr::if_else(
                mets_sum >= 3, "yes", "no", missing = "no"
            )
        ) |>
        # Drop intermediate features calculated to compute metabolic syndrome
        dplyr::select(-dplyr::starts_with("mets_"))

    return(df_result)
}

File: src/feature_engineering.py
import pandas as pd
import numpy as np


def compute_hatch_score(df):
    """
    Computes the HATCH score for each patient based on their age and comorbidities.
    """
    df_result = df.copy()

    # Propagate NaNs to match R's default behavior
    t1 = pd.Series(np.where(df_result["age"] > 75, 1, 0), index=df_result.index).where(
        df_result["age"].notna(), np.nan
    )
    t2 = pd.Series(
        np.where(df_result["hypertension"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["hypertension"].notna(), np.nan)
    t3 = pd.Series(
        np.where(df_result["COPD"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["COPD"].notna(), np.nan)
    t4 = pd.Series(
        np.where(df_result["stroke"] == "yes", 2, 0), index=df_result.index
    ).where(df_result["stroke"].notna(), np.nan)
    t5 = pd.Series(
        np.where(df_result["heart_failure"] == "yes", 2, 0), index=df_result.index
    ).where(df_result["heart_failure"].notna(), np.nan)

    df_result["score_hatch"] = t1 + t2 + t3 + t4 + t5
    return df_result


def compute_chads2_score(df):
    """
    Computes the CHADS2 score for each patient based on their age and comorbidities.
    """
    df_result = df.copy()

    # Propagate NaNs to match R's default behavior
    t1 = pd.Series(np.where(df_result["age"] >= 75, 1, 0), index=df_result.index).where(
        df_result["age"].notna(), np.nan
    )
    t2 = pd.Series(
        np.where(df_result["hypertension"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["hypertension"].notna(), np.nan)
    t3 = pd.Series(
        np.where(df_result["diabetes"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["diabetes"].notna(), np.nan)
    t4 = pd.Series(
        np.where(df_result["stroke"] == "yes", 2, 0), index=df_result.index
    ).where(df_result["stroke"].notna(), np.nan)
    t5 = pd.Series(
        np.where(df_result["heart_failure"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["heart_failure"].notna(), np.nan)

    df_result["score_chads2"] = t1 + t2 + t3 + t4 + t5
    return df_result


def compute_base_af2_score(df):
    """
    Computes the BASE-AF2 score for each patient based on their comorbidities.
    """
    df_result = df.copy()

    # Propagate NaNs to match R's default behavior
    t1 = pd.Series(np.where(df_result["BMI"] >= 28, 1, 0), index=df_result.index).where(
        df_result["BMI"].notna(), np.nan
    )

    cond_la = (
        (df_result["sex"] == "male")
        & df_result["LA_enlargment"].isin(["moderate", "severe"])
    ) | (
        (df_result["sex"] == "female")
        & df_result["LA_enlargment"].isin(["mild", "moderate", "severe"])
    )
    t2 = pd.Series(np.where(cond_la, 1, 0), index=df_result.index).where(
        df_result["sex"].notna() & df_result["LA_enlargment"].notna(), np.nan
    )
    t3 = pd.Series(
        np.where(df_result["smoking_status"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["smoking_status"].notna(), np.nan)
    t4 = pd.Series(
        np.where(df_result["ERAF"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["ERAF"].notna(), np.nan)
    t5 = pd.Series(
        np.where(df_result["AF_duration"] > 6, 1, 0), index=df_result.index
    ).where(df_result["AF_duration"].notna(), np.nan)
    t6 = pd.Series(
        np.where(df_result["AF_type"] == "persistent", 1, 0), index=df_result.index
    ).where(df_result["AF_type"].notna(), np.nan)

    df_result["score_baseaf2"] = t1 + t2 + t3 + t4 + t5 + t6
    return df_result


def compute_mb_later_score(df):
    """
    Computes the MB-LATER score for each patient based on their comorbidities.
    """
    df_result = df.copy()

    # Propagate NaNs to match R's default behavior
    t1 = pd.Series(
        np.where(df_result["sex"] == "male", 1, 0), index=df_result.index
    ).where(df_result["sex"].notna(), np.nan)

    cond_bbb = (df_result["RBBB"] == "yes") | (df_result["LBBB"] == "yes")
    t2 = pd.Series(np.where(cond_bbb, 1, 0), index=df_result.index).where(
        df_result["RBBB"].notna() | df_result["LBBB"].notna(),
        np.nan,
    )

    cond_la_f = (df_result["sex"] == "female") & df_result["LA_enlargment"].isin(
        ["moderate", "severe"]
    )
    t3 = pd.Series(np.where(cond_la_f, 1, 0), index=df_result.index).where(
        df_result["sex"].notna() & df_result["LA_enlargment"].notna(), np.nan
    )

    cond_la_m = (df_result["sex"] == "male") & (df_result["LA_enlargment"] == "severe")
    t4 = pd.Series(np.where(cond_la_m, 1, 0), index=df_result.index).where(
        df_result["sex"].notna() & df_result["LA_enlargment"].notna(), np.nan
    )

    t5 = pd.Series(
        np.where(df_result["AF_type"] == "persistent", 1, 0), index=df_result.index
    ).where(df_result["AF_type"].notna(), np.nan)
    t6 = pd.Series(
        np.where(df_result["ERAF"] == "yes", 1, 0), index=df_result.index
    ).where(df_result["ERAF"].notna(), np.nan)

    df_result["score_mblater"] = t1 + t2 + t3 + t4 + t5 + t6
    return df_result


def compute_metabolic_syndrome(df):
    """
    Computes the metabolic syndrome for each patient based on their comorbidities.
    """
    df_result = df.copy()

    # NAs are treated as 0 (not meeting the condition)
    mets_obesity = np.where(df_result["BMI"] >= 30, 1, 0)

    cond_trig = (df_result["triglycerides"] >= 150) | (df_result["LLA"] == "yes")
    mets_triglycerides = np.where(cond_trig, 1, 0)

    cond_hdl = ((df_result["sex"] == "male") & (df_result["HDL"] < 40)) | (
        (df_result["sex"] == "female") & (df_result["HDL"] < 50)
    )
    mets_hdl = np.where(cond_hdl, 1, 0)

    mets_hypertension = np.where(df_result["hypertension"] == "yes", 1, 0)

    cond_glucose = (df_result["glucose"] >= 100) | (df_result["diabetes"] == "yes")
    mets_glucose = np.where(cond_glucose, 1, 0)

    mets_sum = (
        mets_obesity + mets_triglycerides + mets_hdl + mets_hypertension + mets_glucose
    )

    # metabolic_syndrome is "yes" if mets_sum >= 3, else "no" (missing maps to "no")
    df_result["metabolic_syndrome"] = np.where(mets_sum >= 3, "yes", "no")

    return df_result


def compute_risk_scores(df):
    """
    Computes all risk scores for each patient based on their comorbidities.
    """
    df_result = df.copy()
    df_result = compute_hatch_score(df_result)
    df_result = compute_chads2_score(df_result)
    df_result = compute_base_af2_score(df_result)
    df_result = compute_mb_later_score(df_result)
    df_result = compute_metabolic_syndrome(df_result)

    return df_result
